pageserver/src/basebackup.rs

```diff
@@ -5,180 +5,292 @@
+            slru_buf: [0u8; pg_constants::SLRU_SEG_SIZE],
         }
     }
 
-    // FIXME: Also send all the WAL. The compute node would only need
-    // the WAL that applies to non-relation files, because the page
-    // server handles all the relation files. But we don't have a
-    // mechanism for separating relation and non-relation WAL at the
-    // moment.
-    for entry in std::fs::read_dir(&walpath)? {
-        let entry = entry?;
-        let fullpath = &entry.path();
-        let relpath = fullpath.strip_prefix(&walpath).unwrap();
-
-        if !entry.path().is_file() {
-            continue;
-        }
+	#[rustfmt::skip]
+    pub fn send_tarball(&mut self) -> anyhow::Result<()> {
+        debug!("sending tarball of snapshot in {}", self.snappath);
+        for entry in WalkDir::new(&self.snappath) {
+            let entry = entry?;
+            let fullpath = entry.path();
+            let relpath = entry.path().strip_prefix(&self.snappath).unwrap();
 
-        let archive_fname = relpath.to_str().unwrap();
-        let archive_fname = archive_fname
-            .strip_suffix(".partial")
-            .unwrap_or(&archive_fname);
-        let archive_path = "pg_wal/".to_owned() + archive_fname;
-        ar.append_path_with_name(fullpath, archive_path)?;
-    }
-    ar.finish()?;
-    debug!("all tarred up!");
-    Ok(())
-}
+            if relpath.to_str().unwrap() == "" {
+                continue;
+            }
 
-///
-/// Send a tarball containing a snapshot of all non-relation files in the
-/// PostgreSQL data directory, at given LSN
-///
-/// There must be a snapshot at the given LSN in the snapshots directory, we cannot
-/// reconstruct the state at an arbitrary LSN at the moment.
-///
-pub fn send_snapshot_tarball(
-    write: &mut dyn Write,
-    timelineid: ZTimelineId,
-    snapshotlsn: Lsn,
-) -> Result<(), std::io::Error> {
-    let mut ar = Builder::new(write);
-
-    let snappath = format!("timelines/{}/snapshots/{:016X}", timelineid, snapshotlsn.0);
-    let walpath = format!("timelines/{}/wal", timelineid);
-
-    debug!("sending tarball of snapshot in {}", snappath);
-    //ar.append_dir_all("", &snappath)?;
-
-    for entry in WalkDir::new(&snappath) {
-        let entry = entry?;
-        let fullpath = entry.path();
-        let relpath = entry.path().strip_prefix(&snappath).unwrap();
-
-        if relpath.to_str().unwrap() == "" {
-            continue;
+            if entry.file_type().is_dir() {
+                trace!(
+                    "sending dir {} as {}",
+                    fullpath.display(),
+                    relpath.display()
+                );
+                self.ar.append_dir(relpath, fullpath)?;
+            } else if entry.file_type().is_symlink() {
+                error!("ignoring symlink in snapshot dir");
+            } else if entry.file_type().is_file() {
+                // Shared catalogs are exempt
+                if relpath.starts_with("global/") {
```

Comment:
Are shared catalogs still supposed to be included in the base backup? If the compute node doesn't replay WAL, they will be broken.

Are we missing test coverage for shared catalogs?

Reply:
There is test_createdb which deals with shared catalog. But I agree, more tests need to be added.

Reply:
Shared catalogs are still included in backup. Why? Mostly because it was done in the original implementation and I didn't think much about it. Right now SMGR is handling shared relations in the same way as unshared. So including them in tarball is not mandatory. From the other side, size of shared relation is expected to be small 
so it can be sensible (at least as optimization, reducing node start time) to include them in backup and do not make copute node to request them page-by-page.

Reply:
Yeah, it would be sensible to include them in the backup, and just let the compute node handle them the usual way. I'd like that. However, it won't work unless you also provide all the WAL that applies to the compute node, so that it can recover it. That seems like a non-starter with our current approach.

Reply:
I'd suggest handling shared catalogs in the same way as all other relations.
Now postgres-side logic is simple. If `pageserver_connstr` is set use zenith_smgr for all relations.

We may want to preload some relations on compute node start - see  #214 .
And this mechanism should work for all types of relations too, including catalogs.

Reply:
Pushed